packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -1218,6 +1227,52 @@ function updateTransition(
   return [startTransition, isPending];
 }
 
+let clientId: number = 0;
+
+function mountUniqueId(): string | IdObject {
+  const [id, setId] = mountState(() => {
+    if (getIsHydrating()) {
+      return {
+        $$typeof: REACT_OPAQUE_OBJECT_TYPE,
+        _setId: () => {
+          setId(() => 'c_' + (clientId++).toString());
```

Comment:
This has the same problem as most of the "userspace" solutions: this ID is unstable and depends on commit order, especially in combination with Suspense and partial hydration.

One thing library authors often asked for is a way of getting the "tree position" of the element, which is better and relatively stable, but still not perfect; it'll probably work better for hydration and will be globally unique, but might get reused if a component is swapped by another for example.

Reply:
I think with this solution it doesn't matter whether the counter changes between server and client since react is able to patch it up properly during hydration which wasn't possible before.

This hook is exclusively for IDREF attributes. It's not a generic, unique ID of your component instance.

Reply:
Ah, so the point is it not being seen as a hydration error but just as something to be patched 🤔

Reply:
At least this is how I understood it. There's an argument to be made about hydration warnings for attributes in the first place (e.g. `<button disabled={!process.browser} />` but IDREFs seem like a safe place to start.

Reply:
It’s not as simple as just patching. Partially because we don’t actually compare any props in hydration. But mostly because if you do it the naive way you may patch up one side but not the other side when partial hydration is being used.

This enforces that we force hydration of the things that needs patching on both sides.

We should have a test for that.